From: Michael Grzeschik <m.grzeschik@pengutronix.de>
Date: Thu, 20 Sep 2018 09:08:03 +0200
Subject: [PATCH] dsa: add self_test ethtool function handling

This patch adds the support for running the self_test
callback on dsa based interfaces.

Signed-off-by: Michael Grzeschik <m.grzeschik@pengutronix.de>
---
 include/net/dsa.h |  2 ++
 net/dsa/slave.c   | 20 +++++++++++++++++---
 2 files changed, 19 insertions(+), 3 deletions(-)

diff --git a/include/net/dsa.h b/include/net/dsa.h
index a7ed97629969..fd81b950048c 100644
--- a/include/net/dsa.h
+++ b/include/net/dsa.h
@@ -408,6 +408,8 @@ struct dsa_switch_ops {
 	void	(*get_ethtool_phy_stats)(struct dsa_switch *ds,
 					 int port, uint64_t *data);
 
+	void	(*self_test)(struct dsa_switch *ds, int port,
+			     struct ethtool_test *test, uint64_t *data);
 	/*
 	 * ethtool Wake-on-LAN
 	 */
diff --git a/net/dsa/slave.c b/net/dsa/slave.c
index f734ce0bcb56..a8aeda677031 100644
--- a/net/dsa/slave.c
+++ b/net/dsa/slave.c
@@ -655,6 +655,17 @@ static void dsa_slave_get_strings(struct net_device *dev,
 	}
 }
 
+static void dsa_slave_self_test(struct net_device *dev,
+				struct ethtool_test *test,
+				uint64_t *data)
+{
+	struct dsa_port *dp = dsa_slave_to_port(dev);
+	struct dsa_switch *ds = dp->ds;
+
+	if (ds->ops->self_test)
+		ds->ops->self_test(ds, dp->index, test, data);
+}
+
 static void dsa_slave_get_ethtool_stats(struct net_device *dev,
 					struct ethtool_stats *stats,
 					uint64_t *data)
@@ -690,11 +701,13 @@ static int dsa_slave_get_sset_count(struct net_device *dev, int sset)
 {
 	struct dsa_port *dp = dsa_slave_to_port(dev);
 	struct dsa_switch *ds = dp->ds;
+	int count = 0;
 
-	if (sset == ETH_SS_STATS) {
-		int count;
-
+	switch (sset) {
+	case ETH_SS_STATS:
 		count = 4;
+		/* fall through */
+	case ETH_SS_TEST:
 		if (ds->ops->get_sset_count)
 			count += ds->ops->get_sset_count(ds, dp->index, sset);
 
@@ -1191,6 +1204,7 @@ static const struct ethtool_ops dsa_slave_ethtool_ops = {
 	.get_rxnfc		= dsa_slave_get_rxnfc,
 	.set_rxnfc		= dsa_slave_set_rxnfc,
 	.get_ts_info		= dsa_slave_get_ts_info,
+	.self_test		= dsa_slave_self_test,
 };
 
 /* legacy way, bypassing the bridge *****************************************/
